router/core/cache_warmup.go

```diff
@@ -0,0 +1,278 @@
+}
+
+func (w *cacheWarmup) run(ctx context.Context) (int, error) {
+
+	ctx, cancel := context.WithTimeout(ctx, w.timeout)
+	defer cancel()
+
+	items, err := w.source.LoadItems(ctx, w.log)
+	if err != nil {
+		return 0, err
+	}
+
+	if len(items) == 0 {
+		w.log.Info("Cache warmup - no items to process")
+		return 0, nil
+	}
+
+	w.log.Info("Cache warmup - items loaded, starting processing",
+		zap.Int("items", len(items)),
+	)
+
+	defaultClientInfo := &ClientInfo{}
+
+	done := ctx.Done()
+	index := make(chan int, len(items))
+	defer close(index)
+	itemCompleted := make(chan struct{})
+
+	for i, item := range items {
+		if item.Client == nil {
+			item.Client = defaultClientInfo
+		}
+		index <- i
+	}
+
+	var (
+		rl ratelimit.Limiter
+	)
+
+	if w.itemsPerSecond > 0 {
+		rl = ratelimit.New(w.itemsPerSecond)
+	} else {
+		rl = ratelimit.NewUnlimited()
+	}
+
+	for i := 0; i < w.workers; i++ {
+		go func(i int) {
+			for {
+				select {
+				case <-done:
+					return
+				case idx, ok := <-index:
+					if !ok {
+						return
+					}
+					rl.Take()
+					item := items[idx]
+					err := w.processOperation(ctx, item)
+					if err != nil {
+						w.log.Error("Failed to process operation, skipping",
+							zap.Error(err),
+							zap.String("clientName", item.Client.Name),
+							zap.String("clientVersion", item.Client.Version),
+							zap.String("query", item.Request.Query),
+							zap.String("operationName", item.Request.OperationName),
+						)
+					}
+					select {
+					case <-done:
+						return
+					case itemCompleted <- struct{}{}:
+					}
+				}
+			}
+		}(i)
+	}
+
+	for i := 0; i < len(items); i++ {
+		select {
+		case <-done:
+			return i, ctx.Err()
+		case <-itemCompleted:
+			processed := i + 1
+			if processed%100 == 0 {
+				w.log.Info("Cache warmup - processed items",
+					zap.Int("processedItems", processed),
+				)
+			}
+		}
+	}
+
+	return len(items), nil
+}
+
+func (w *cacheWarmup) processOperation(ctx context.Context, item *CacheWarmupItem) error {
+
+	var (
+		isAPQ bool
+	)
+
+	k, err := w.operationProcessor.NewIndependentKit()
+	if err != nil {
+		return err
+	}
+
+	k.parsedOperation.Request = item.Request
+
+	err = k.unmarshalOperation()
+	if err != nil {
+		return err
+	}
+
+	err = k.ComputeOperationSha256()
+	if err != nil {
+		return err
+	}
+
+	if k.parsedOperation.IsPersistedOperation {
+		_, isAPQ, err = k.FetchPersistedOperation(ctx, item.Client)
+		if err != nil {
+			return err
+		}
+	}
+
+	err = k.Parse()
+	if err != nil {
+		return err
+	}
+
+	_, err = k.NormalizeOperation(item.Client.Name, isAPQ)
+	if err != nil {
+		return err
+	}
+
+	err = k.NormalizeVariables()
+	if err != nil {
+		return err
+	}
+
+	_, err = k.Validate(true)
+	if err != nil {
+		return err
+	}
+
+	if w.complexityLimits != nil {
+		_, _, _ = k.ValidateQueryComplexity(w.complexityLimits, k.kit.doc, w.routerSchema, k.parsedOperation.IsPersistedOperation)
+	}
+
+	planOptions := PlanOptions{
+		Protocol:   OperationProtocolHTTP,
```

Comment:
This won't work with subscriptions.

Reply:
We didn't use this at all so I've removed it.